File: source/ts/src/astronomy/sunrise.ts
import { PanchangError } from '../types/errors';
import { validateLocation } from '../utils/validation';
import { resolveEvent, type RiseSetKind } from './riseSetCache';
import type { GeoLocation } from '../types/location';

const SOLAR: RiseSetKind = { body: 'sun' };

/**
 * First sunrise at or after the given UTC search start; throws `NO_SUNRISE` under midnight sun or polar night.
 * `limitDays` defaults to 2 (the Go port has no default, so there its zero value searches nothing). An Invalid Date,
 * or a search that would reach 2^52 ms from 1970 where the solver cannot resolve a millisecond, throws `INVALID_DATE`.
 */
export function computeSunrise(
  searchFromUtc: Date,
  location: GeoLocation,
  limitDays: number = 2
): Date {
  validateLocation(location);
  const result = resolveEvent(SOLAR, +1, searchFromUtc, location, limitDays);

  if (!result) {
    throw new PanchangError(
      `No sunrise found within ${limitDays} days for ` +
        `(${location.latitude}°, ${location.longitude}°) near ${searchFromUtc.toISOString()}. ` +
        `This location may be experiencing midnight sun or polar night.`,
      'NO_SUNRISE'
    );
  }

  return result;
}

/**
 * First sunset at or after the given UTC search start; throws `NO_SUNSET` under midnight sun or polar night.
 * `limitDays` defaults to 2 (the Go port has no default). An Invalid Date, or a search that would reach 2^52 ms from
 * 1970, throws `INVALID_DATE`.
 */
export function computeSunset(
  searchFromUtc: Date,
  location: GeoLocation,
  limitDays: number = 2
): Date {
  validateLocation(location);
  const result = resolveEvent(SOLAR, -1, searchFromUtc, location, limitDays);

  if (!result) {
    throw new PanchangError(
      `No sunset found within ${limitDays} days for ` +
        `(${location.latitude}°, ${location.longitude}°) near ${searchFromUtc.toISOString()}.`,
      'NO_SUNSET'
    );
  }

  return result;
}

File: source/go/internal/astronomy/sunrise.go
package astronomy

import (
	"time"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var solar = RiseSetKind{Body: RiseSetSun}

const DefaultRiseSetLimitDays = 2

func ComputeSunrise(ctx *EphemerisCtx, searchFromMs int64, location types.GeoLocation, limitDays int) (int64, error) {
	if err := utils.ValidateLocation(location); err != nil {
		return 0, err
	}
	result, ok, err := ResolveEvent(ctx, solar, +1, searchFromMs, location, limitDays)
	if err != nil {
		return 0, err
	}
	if !ok {
		return 0, types.Codef(types.ErrNoSunrise,
			"No sunrise found within %d days for (%s°, %s°) near %s. "+
				"This location may be experiencing midnight sun or polar night.",
			limitDays, jsnum.FormatFloat(location.Latitude), jsnum.FormatFloat(location.Longitude),
			types.JSDate(searchFromMs).ISOString())
	}
	return result, nil
}

func ComputeSunset(ctx *EphemerisCtx, searchFromMs int64, location types.GeoLocation, limitDays int) (int64, error) {
	if err := utils.ValidateLocation(location); err != nil {
		return 0, err
	}
	result, ok, err := ResolveEvent(ctx, solar, -1, searchFromMs, location, limitDays)
	if err != nil {
		return 0, err
	}
	if !ok {
		return 0, types.Codef(types.ErrNoSunset,
			"No sunset found within %d days for (%s°, %s°) near %s.",
			limitDays, jsnum.FormatFloat(location.Latitude), jsnum.FormatFloat(location.Longitude),
			types.JSDate(searchFromMs).ISOString())
	}
	return result, nil
}

func isoString(ms int64) string {
	return time.UnixMilli(ms).UTC().Format("2006-01-02T15:04:05.000Z")
}
